### DISCUSSION TITLE
block access to application after authenticated with external IDP

### DISCUSSION THREAD

**Discussion by kluangguy1** [keycloak/keycloak#13743] [Q&A] [answered]
Dear all,
I have a use case below, could I know is it something achievable in Keycloak?

User A trying to access application E.G inventory application, once user A is authenticated with IDP E.G OpenID 2.0 / SAML 2.0, user A will be blocked to access inventory application based on user name E.G s=userA

 "sub": "s=userA,u=8e191461-6a45-4839-adc3-7e0c0a819b6a",

I have tried by defining a new flow part of authentication with below
1) Condition - User Attribute
2) Deny Access

Basically the condition will be checking whether the username is matching "userA", if so, it will block the access. however it is not functioning as my intention, the keycloak version is v18.0

Any input/idea to achieve the use case will be greatly appreciated, thanks!

**Comment by bonnm** [comment, marked as answer]
Hello,
After external IDP authentication, flows are not continued, execution stops:
[https://github.com/keycloak/keycloak/issues/10250](https://github.com/keycloak/keycloak/issues/10250)
This is an age-old bug which never was corrected afaik.
You can try to put your conditional auth flow in a post-broker-login flow with the drawback having it beeing executed every time for every client...

**Reply by kluangguy1** [reply]
Thanks, by putting under post-broker-login it works

**Comment by cgeorgilakis** [comment]
For us also, this Keycloak bug is blocking for our usecase because it breaks the ability to use Keycloak as Service Provider and Identity Brokering together. Client Authentication flow must executed in the same way for Keycloak Users and external IdPs Users.

We have created [a PR](https://github.com/keycloak/keycloak/pull/12891) for this. Keycloak team could you help as overcome test failures and review this PR? This PR has been checked successfully for basic use cases. See details in [the PR](https://github.com/keycloak/keycloak/pull/12891).

**Reply by sschu** [reply]
@cgeorgilakis Why can't you continue your flow in a post-broker login flow?

**Comment by anand-kulk** [comment]
You should look into if you can use Authorizations tab in client settings (enable fine grained authorizations). There you can define resources, policies, permissions. You could create a user based policy and apply permission to a resource. The effective permissions to resources get populated in the UMA token which you can exchange for access token. Essentially, you use external IDP to check authentication and use authorization to control access.

**Comment by cgeorgilakis** [comment]
@sschu IdP post-login flow can not replace the need for proper client browser flow execution for external IdP users.
 IdP post-login flow is executed after specific successful IdP login. 
Client browser flow is executed per client no matter from where user is logged in ( Keycloak User, from which external IdP). Different clients need to have different authentication method  ( see examples in [github issue](https://github.com/keycloak/keycloak/issues/10250#issuecomment-1222384798) ). That's why Keycloak must continue Client browser flow after successful IdP login.

**Reply by sschu** [reply]
@cgeorgilakis Maybe you can achieve what you want by using step-up authentication? You can configure default ACR values per client in the advanced client settings.

**Reply by cgeorgilakis** [reply]
@sschu we have tested it. Unfortunately, due to this bug neither step-up authentication works. Step-up authentication needs an appropriate client browser flow to be configured.
step-up authentication is one of the needed client feature for us.

**Reply by sschu** [reply]
I see.
